src/persist/src/nemesis/progress.rs

```diff
@@ -0,0 +1,172 @@
+    /// `!dataflow.less_equal(ts)`).
+    pub fn wait(&self, ts: u64) -> Future<()> {
+        let (tx, rx) = Future::new();
+        if self.closed.load(Ordering::SeqCst) {
+            tx.fill(Err(Error::RuntimeShutdown));
+            return rx;
+        }
+        if let Err(err) = self.tx.send((ts, tx)) {
+            let (_, tx) = err.into_inner();
+            tx.fill(Err(Error::RuntimeShutdown));
+        }
+        rx
+    }
+}
+
+/// The dataflow handle for a [DataflowProgress].
+#[derive(Clone, Debug)]
+pub struct DataflowProgressHandle {
+    core: Arc<Mutex<DataflowProgressCore>>,
+}
+
+impl DataflowProgressHandle {
+    /// Advance the internal frontier and unblock any relevant waiters.
+    ///
+    /// This is intended to be called in the dataflow step loop:
+    ///
+    /// ```
+    /// let (progress_tx, progress_rx) = DataflowProgress::new();
+    /// ...
+    /// timely::execute(
+    ///     timely::Config::process(workers),
+    ///     move |worker| {
+    ///         let dataflow_read = dataflow_read.clone();
+    ///         let mut probe = ProbeHandle::new();
+    ///         worker.dataflow(|scope| {
+    ///             let out = ...;
+    ///             out.probe_with(&mut probe);
+    ///         })
+    ///         while worker.step_or_park(None) {
+    ///             progress_tx.maybe_progress(&probe);
+    ///         }
+    ///         progress_tx.close();
+    ///     }
+    /// )
+    /// ```
+    pub fn maybe_progress(&self, probe: &ProbeHandle<u64>) {
+        if let Ok(mut core) = self.core.lock() {
+            core.maybe_progress(probe)
+        }
+    }
+
+    /// Close down this [DataflowProgress] and fill any remaining waiters with an
+    /// error.
+    pub fn close(&self) {
+        if let Ok(mut core) = self.core.lock() {
+            core.close()
+        }
+    }
+}
+
+#[derive(Debug)]
+struct DataflowProgressCore {
+    // See the field documentation in DataflowProgress for the protocol on using
+    // these.
+    closed: Arc<AtomicBool>,
+    rx: crossbeam_channel::Receiver<(u64, FutureHandle<()>)>,
+
+    waiters: BTreeMap<u64, Vec<FutureHandle<()>>>,
+    frontier: Antichain<u64>,
+}
+
+impl DataflowProgressCore {
+    fn maybe_progress(&mut self, probe: &ProbeHandle<u64>) {
+        self.drain_rx();
+        probe.with_frontier(|frontier| {
+            if PartialOrder::less_than(&self.frontier.borrow(), &frontier) {
+                self.frontier.clone_from(&frontier.to_owned());
+                let mut removed = Vec::new();
+                for (wait_ts, waiters) in self.waiters.iter_mut() {
+                    if !frontier.less_than(wait_ts) {
+                        removed.push(*wait_ts);
+                        waiters.drain(..).for_each(|w| {
+                            w.fill(Ok(()));
+                        })
+                    } else {
+                        // If the timestamps are totally ordered, the BTreeMap
```

Comment:
Very minor nit: I would phrase that as "Since the timestamps are totally ordered", because the timestamps are always u64 in the current code?

Reply:
Done